src/Microsoft.AspNet.OData.Shared/Formatter/ODataOutputFormatterHelper.cs

```diff
@@ -349,5 +348,65 @@ private static bool IsOperationPath(ODataPath path)
 
             return false;
         }
+
+        private static Microsoft.OData.UriParser.ODataPath ConvertPath(ODataPath path)
+        {
+            if (path == null)
+            {
+                return null;
+            }
+
+            if (IsOperationPath(path))
+            {
+                var lastSegment = path.Segments.Last();
+                OperationSegment operation = lastSegment as OperationSegment;
+                if (operation != null && operation.EntitySet != null)
+                {
+                    return GeneratePath(operation.EntitySet);
+                }
+
+                OperationImportSegment operationImport = lastSegment as OperationImportSegment;
+                if (operationImport != null && operationImport.EntitySet != null)
+                {
+                    return GeneratePath(operationImport.EntitySet);
+                }
+
+                return null;
+            }
+
+            return path.Path;
+        }
+
+        private static Microsoft.OData.UriParser.ODataPath GeneratePath(IEdmNavigationSource navigationSource)
+        {
+            switch (navigationSource.NavigationSourceKind())
+            {
+                case EdmNavigationSourceKind.EntitySet:
+                    return new Microsoft.OData.UriParser.ODataPath(new EntitySetSegment((IEdmEntitySet)navigationSource));
+
+                case EdmNavigationSourceKind.Singleton:
+                    return new Microsoft.OData.UriParser.ODataPath(new SingletonSegment((IEdmSingleton)navigationSource));
+
+                case EdmNavigationSourceKind.ContainedEntitySet:
+                    IEdmContainedEntitySet containedEntitySet = (IEdmContainedEntitySet)navigationSource;
+                    var path = GeneratePath(containedEntitySet.ParentNavigationSource);
+                    IList<ODataPathSegment> segments = new List<ODataPathSegment>();
+                    if (path != null)
+                    {
+                        foreach (var item in path)
+                        {
+                            segments.Add(item);
+                        }
+                    }
+
+                    segments.Add(new NavigationPropertySegment(containedEntitySet.NavigationProperty, containedEntitySet.ParentNavigationSource));
```

Comment:
is this the right behavior for path == null?  if path==null, shouldn't you just return null?

Reply:
Emm... probably.  in ODL, IEdmContainedEntitySet interface is public, but there's only one implementation "EdmContainedEntitySet" is internal. That means user can't create a "EdmContainedEntitySet" instance. If you can get an instance of "IEdmContainedEntitySet", the ParentNavigationSource should be valid. So, i'd like to remove the conditional verification. 

---
In reply to: [386708095](https://github.com/OData/WebApi/pull/2060#discussion_r386708095) [](ancestors = 386708095)

Reply:
This method looks similar to the ComputePath() method in EdmContainedEntitySet, which is unfortunately internal.  However, that method deals with nav props defined on derived types. Do we need to do anything to handle that case here?

Also a bit concerned that EdmContainedEntitySet allows passing in a navigationPath, which is used in ComputePath. What are the situations in which that would be something other than the navigationProperty? Would this be used for nav props on complex types? Do we need to handle that case here?

---
In reply to: [386750791](https://github.com/OData/WebApi/pull/2060#discussion_r386750791) [](ancestors = 386750791,386708095)

Reply:
Consider adding an assert that containedEntitySet.ParentNavigationSource != null.

---
In reply to: [386750791](https://github.com/OData/WebApi/pull/2060#discussion_r386750791) [](ancestors = 386750791,386708095)

Reply:
EdmContainedEntitySet constructor makes the input check. That can make sure "ParentNavigationSource" never be null if the "EntityContainedEntitySet" is live.

public EdmContainedEntitySet(IEdmNavigationSource parentNavigationSource, IEdmNavigationProperty navigationProperty, IEdmPathExpression navigationPath)
            : base(navigationProperty.Name, navigationProperty.ToEntityType())
        {
            EdmUtil.CheckArgumentNull(parentNavigationSource, "parentNavigationSource");
            ...

            this.parentNavigationSource = parentNavigationSource;
            this.navigationProperty = navigationProperty;
            this.navigationPath = navigationPath;
        }


---
In reply to: [391139997](https://github.com/OData/WebApi/pull/2060#discussion_r391139997) [](ancestors = 391139997,386750791,386708095)

Reply:
Besides, for the  ComputePath() method in EdmContainedEntitySet, we can call " containedEntitySet.Path" to get the Path. However, "Path" is IEdmPathExpression, it's a string segment.
We want the ODataPath, it's a path segments. So, we have to use "OData Uri parse" to parse the "IEdmPathExpression". That's not feasible here. 
I'd like to create a new issue to track on this feature request.  And I do think that's a feature in the ODL . 

@mikep

---
In reply to: [394725483](https://github.com/OData/WebApi/pull/2060#discussion_r394725483) [](ancestors = 394725483,391139997,386750791,386708095)

Reply:
@saxu@microsoft.com -- did you create an issue to track this?  Please specifically call out what is not currently supported (i.e., how does this code handle navigation properties on complex types or navigation properties defined on derived types? does it work with navigation properties on entities returned from contained navigation properties?)

---
In reply to: [394726333](https://github.com/OData/WebApi/pull/2060#discussion_r394726333) [](ancestors = 394726333,394725483,391139997,386750791,386708095)